### ISSUE TITLE
mksquashfs number of processers should be passable from the environment

### ISSUE THREAD

**Issue by rhpvorderman** [apptainer/singularity#4397]
### Expected behavior
We run singularity on a compute cluster. For each job we request a certain amount of memory and cores based on our expected usage.

Singularity exec docker://quay.io/biocontainers/samtools:some_tag samtools flagstat my.bam should only take 1 core.

### Actual behavior

When the image is not yet present it will be downloaded from the docker hub (quay.io in this case). The image will be build. This uses mksquashfs. By default mksquashfs **uses all cores available on the system.** On our compute nodes that is usually 32. As a result this simple job will use way much more memory and CPU than anticipated and the job will be killed on the cluster.

### Steps to reproduce behavior
singularity exec docker://anyimageyoudonothaveyetpreferablyabigone echo "succes!"

Will see all your system's cpu resources used.

### Proposed solution
Add a SINGULARITY_MKSQUASHFS_OPTS env variable or a SINGULARITY_MKSQUASHFS_PROCESSORS or a SINGULARITY_BUILD_PROCESSORS variable.
This way it is very easy to do `export SINGULARITY_MKSQUASHFS_OPTS=-processors 1` or  `export SINGULARITY_BUILD_PROCESSORS=1`.

**Comment by cclerget** [comment]
Hi @rhpvorderman , `taskset -c 0 singularity exec docker://...` doesn't help ?

**Comment by cclerget** [comment]
Nevermind for taskset, stupid suggestion for a HPC job :). Between Is 2.3.1 or 3.2.1 version ?

**Comment by rhpvorderman** [comment]
whoops 3.2.1. Will change.

**Comment by cclerget** [comment]
Ok as a possible workaround :

```
taskset -c 0 singularity exec docker://... true
singularity exec docker://...
```

The first command downloads/builds/caches image and the second could use the SIF image from the cache

**Comment by rhpvorderman** [comment]
> Ok as a possible workaround :
> 
> ```
> taskset -c 0 singularity exec docker://... true
> singularity exec docker://...
> ```
> 
> The first command downloads/builds/caches image and the second could use the SIF image from the cache

Taskset is not a very nice thing to do on a hpc worker node, as you claim a certain CPU core. But the idea of running the caching in a separate step is very good. Maybe there are other tools around such as cpulimit that can help here.

Thank you for the suggestion.

**Comment by rhpvorderman** [comment]
As a follow up. I used cpulimit (can be installed from Ubuntu and Debian repo's). This helped in the CPU usage area.
Unfortunately this does not help in the memory usage area. All these threads get allocated a certain amount of memory and our cluster (unfortunately) kills on vmem. That goes from 300MB (no problem) on one thread to 4800MB (exceeds the default) using all threads.
It would be great if the number of processors used by mksquashfs could be a setting in a config/environment somewhere.

**Comment by dtrudg** [comment]
Closed in #5003
